sklearn/metrics/classification.py

```diff
@@ -391,72 +391,269 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
+        n_labels = None
+    else:
+        n_labels = len(labels)
+        labels = np.hstack([labels, np.setdiff1d(present_labels, labels,
+                                                 assume_unique=True)])
+
     if y_type.startswith('multilabel'):
+        if average in (None, 'samples'):
+            if normalize is None:
+                # default is average='samples'
+                average = 'samples'
+            else:
+                if normalize:
+                    average = 'samples'
+                else:
+                    average = 'none-samples'
+                warn_message = ("'normalize' was removed in version 0.20 and "
+                                "will be removed in 0.22, instead use "
+                                "`average='%s'`." % normalize)
+                warnings.warn(warn_message, DeprecationWarning)
+        # else:
+            # otherwise what should we do? raise warning or ValueError?
+
+        if not np.all(labels == present_labels):
+            if np.max(labels) > np.max(present_labels):
+                raise ValueError('All labels must be in [0, n, labels). '
+                                 'Got %d > %d' %
+                                 (np.max(labels), np.max(present_labels)))
+            if np.min(labels) < 0:
+                raise ValueError('All labels must be in [0, n, labels). '
+                                 'Got %d < 0' % np.min(labels))
+
+        # wait for response on 'prf-bug' PR, since I'm less than 90% sure
+        if n_labels is not None:
+            y_true = y_true[:, labels[:n_labels]]
+            y_pred = y_pred[:, labels[:n_labels]]
+
         with np.errstate(divide='ignore', invalid='ignore'):
-            # oddly, we may get an "invalid" rather than a "divide" error here
-            pred_or_true = count_nonzero(y_true + y_pred, axis=1)
-            pred_and_true = count_nonzero(y_true.multiply(y_pred), axis=1)
-            score = pred_and_true / pred_or_true
-            score[pred_or_true == 0.0] = 1.0
+
+            if average == 'samples':
+                pred_or_true = count_nonzero(y_true + y_pred, axis=1)
+                pred_and_true = count_nonzero(y_true.multiply(y_pred),
+                                              axis=1)
+                score = pred_and_true / pred_or_true
+                score[pred_or_true == 0.0] == 1.0
+                return _weighted_sum(score, sample_weight, normalize=True)
+            elif average == 'none-samples':
+                pred_or_true = count_nonzero(y_true + y_pred, axis=1)
+                pred_and_true = count_nonzero(y_true.multiply(y_pred),
+                                              axis=1)
+                score = pred_and_true / pred_or_true
+                score[pred_or_true == 0.0] == 1.0
+                return _weighted_sum(score, sample_weight, normalize=False)
+            elif average == 'micro':
+                pred_or_true = count_nonzero(y_true + y_pred, axis=1,
```

Comment:
You should be able to handle the entire multilabel case with just two calls to count_nonzero in the code

Reply:
Done.